stdlib/type.mli

```diff
@@ -35,3 +35,78 @@ type (_, _) eq = Equal: ('a, 'a) eq
 
     At runtime, this function simply returns its second argument unchanged.
 *)
+
+(** {1:identifiers Type identifiers} *)
+
+(** Type identifiers.
+
+    Type identifiers are values that can be tested for
+    {{!Id.equal}equality} to prove they represent the same type. Note
+    that unequal identifiers does not imply unequal types: the same
+    type can be represented by multiple identifiers.
+
+    See an {{!Id.example}example} of use. *)
+module Id : sig
+
+  (** {1:ids Type identifiers} *)
+
+  type 'a t
+  (** The type for type identifiers. ['a] is the represented type. *)
+
+  val make : unit -> 'a t
+  (** [make ()] is a new type identifier. *)
+
+  val equal : 'a t -> 'b t -> ('a, 'b) eq option
+  (** [equal id0 id1] is [Some Equal] if identifier [id0] is equal to [id1]
+      and [None] otherwise. *)
+
+  (** {1:example Example}
+
+      The following shows how type identifiers can be used to implement
+      an heterogeneous key-value dictionary.
+{[
+(** Heterogeneous dictionaries. *)
+module Dict : sig
+  type t
+  (** The type for dictionaries. *)
+
+  type 'a key
+  (** The type for keys binding values of type ['a]. *)
+
+  val key : unit -> 'a key
+  (** [key ()] is a new dictionary key. *)
+
+  val empty : t
+  (** [empty] is the empty dictionary. *)
+
+  val add : 'a key -> 'a -> t -> t
+  (** [add k v d] is [d] with [k] bound to [v]. *)
+
+  val remove : 'a key -> t -> t
+  (** [remove k d] is [d] with the last binding of [k] removed. *)
+
+  val find : 'a key -> t -> 'a option
+  (** [find k d] is the binding of [k] in [d], if any. *)
+end = struct
+  type 'a key = { tid : 'a Type.Id.t; exist : exist_key }
+  and exist_key = K : 'a key -> exist_key
+
+  let key () = let rec k = { tid = Type.Id.make (); exist = K k } in k
+
+  type binding = B : 'a key * 'a -> binding
```

Comment:
Slight simplification (if the one I suggested above isn't applied):

```suggestion
  type binding = B : 'a Type.Id.t * 'a -> binding
```

(construct `B (k.tid, v)` in `add`, use the matched `tid` directly in `find`)

Reply:
I prefer to keep the full key here. This makes the example easier to tweak with more interesting functionality like augmenting `'a key` with a name and a pretty-printer which in turns allows to provide a formatter for dictionaries.